File: cs25backend/src/main/java/edu/dongguk/cs25backend/service/ManagerService.java
package edu.dongguk.cs25backend.service;

import edu.dongguk.cs25backend.domain.Manager;
import edu.dongguk.cs25backend.dto.request.ManagerDto;
import edu.dongguk.cs25backend.exception.CS25Exception;
import edu.dongguk.cs25backend.exception.ErrorCode;
import edu.dongguk.cs25backend.repository.ManagerRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
@Transactional
@RequiredArgsConstructor
public class ManagerService {
    private final ManagerRepository managerRepository;

    //C
    public Boolean createManager(ManagerDto dto) {
        managerRepository.findByLoginIdOrNameOrEmailOrPhoneNumber(dto.getLoginId())
                .ifPresent(m -> {
                    throw new CS25Exception(ErrorCode.DUPLICATION_MANAGER);
                });

        managerRepository.save(Manager.builder()
                .loginId(dto.getLoginId())
                .password(dto.getPassword())
                .name(dto.getName())
                .email(dto.getEmail())
                .phoneNumber(dto.getPhoneNumber())
                .userRole(dto.getUserRole())
                .memberShip(dto.getMemberShip())
                .build());

        return Boolean.TRUE;
    }
    //R
    //보류

    //U
    //보류

    //D
    //디비에서 지우죠!

}

File: cs25server/src/main/kotlin/edu/dongguk/cs25server/service/ManagerService.kt
package edu.dongguk.cs25server.service


import edu.dongguk.cs25server.domain.Manager
import edu.dongguk.cs25server.domain.type.AllowStatus
import edu.dongguk.cs25server.dto.request.ManagerRequestDto
import edu.dongguk.cs25server.dto.response.ListResponseDto
import edu.dongguk.cs25server.dto.response.RequestManagerListDto
import edu.dongguk.cs25server.dto.response.PageInfo
import edu.dongguk.cs25server.exception.ErrorCode
import edu.dongguk.cs25server.exception.GlobalException
import edu.dongguk.cs25server.repository.ManagerRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class ManagerService(
    private val managerRepository: ManagerRepository,
) {
    fun createManager(requestDto: ManagerRequestDto): Boolean {
        managerRepository.findTop1ByLoginIdOrNameOrEmailOrPhoneNumber(
            requestDto.loginId,
            requestDto.name,
            requestDto.email,
            requestDto.phoneNumber
        )?.let {
            throw GlobalException(ErrorCode.DUPLICATION_MANAGER)
        }
        managerRepository.save(
            Manager(
                loginId = requestDto.loginId,
                password = requestDto.password,
                name = requestDto.name,
                email = requestDto.email,
                phoneNumber = requestDto.phoneNumber,
                role = requestDto.userRole,
                memberShip = requestDto.memberShip
            )
        )
        return true
    }

    fun getRequestManagerList(pageIndex: Long, pageSize: Long): ListResponseDto<List<RequestManagerListDto>> {
        val paging: Pageable = PageRequest.of(
            pageIndex.toInt(),
            pageSize.toInt(),
        )

        val managerList: Page<ManagerRepository.ManagerInfo> =
            managerRepository.findNotAllowManagerByStatusOrderByCreatedAtDesc(AllowStatus.BEFORE.toString(), paging);

        val pageInfo: PageInfo = PageInfo(
            page = pageIndex.toInt(),
            size = pageSize.toInt(),
            totalElements = managerList.totalElements.toInt(),
            totalPages = managerList.totalPages
        )

        val managerDtoList: List<RequestManagerListDto> = managerList
            .map { m ->
                RequestManagerListDto(
                    m.getID(),
                    m.getNAME(),
                    m.getPNUMBER(),
                    m.getADDRESS(),
                    m.getCREATEDAT().toString()
                )
            }
            .toList()

        return ListResponseDto(managerDtoList, pageInfo)
    }

    fun applyManagerRequest(managerId: Long, select: Boolean): Boolean {
        val manager = managerRepository.findByIdAndStatus(managerId, AllowStatus.BEFORE)
            ?: throw GlobalException(ErrorCode.NOT_FOUND_MANAGER)

        if (select) {
            manager.allowManager(AllowStatus.APPROVAL)
        } else {
            manager.allowManager(AllowStatus.REFUSE)
        }

        return true;
    }
}


